protocol/swap/manager.go

```
@@ -1,168 +1,156 @@
 
-	switch info.status.IsOngoing() {
+	switch info.Status.IsOngoing() {
 	case true:
-		m.ongoing[info.id] = info
+		m.ongoing[info.ID] = info
 	default:
-		m.past[info.id] = info
+		m.past[info.ID] = info
 	}
 
-	return nil
+	return m.db.PutSwap(info)
 }
 
 // GetPastIDs returns all past swap IDs.
-func (m *manager) GetPastIDs() []types.Hash {
+func (m *manager) GetPastIDs() ([]types.Hash, error) {
 	m.RLock()
 	defer m.RUnlock()
-	ids := make([]types.Hash, len(m.past))
-	i := 0
+	ids := make(map[types.Hash]struct{})
 	for id := range m.past {
-		ids[i] = id
+		ids[id] = struct{}{}
+	}
+
+	// TODO: do we want to cache all past swaps since we're already fetching them?
+	stored, err := m.db.GetAllSwaps()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range stored {
+		if s.Status.IsOngoing() {
+			continue
+		}
+
+		ids[s.ID] = struct{}{}
+	}
+
+	idArr := make([]types.Hash, len(ids))
+	i := 0
+	for id := range ids {
+		idArr[i] = id
 		i++
 	}
-	return ids
+
+	return idArr, nil
 }
 
 // GetPastSwap returns a swap's *Info given its ID.
-func (m *manager) GetPastSwap(id types.Hash) *Info {
+func (m *manager) GetPastSwap(id types.Hash) (*Info, error) {
 	m.RLock()
 	defer m.RUnlock()
-	return m.past[id]
+	s, has := m.past[id]
+	if has {
+		return s, nil
+	}
+
+	// TODO: do we want to cache this swap?
+	return m.getSwapFromDB(id)
 }
 
 // GetOngoingSwap returns the ongoing swap's *Info, if there is one.
-func (m *manager) GetOngoingSwap(id types.Hash) *Info {
+func (m *manager) GetOngoingSwap(id types.Hash) (*Info, error) {
 	m.RLock()
 	defer m.RUnlock()
-	return m.ongoing[id]
+	s, has := m.ongoing[id]
+	if !has {
+		return nil, errNoSwapWithID
+	}
+
+	return s, nil
 }
 
 // CompleteOngoingSwap marks the current ongoing swap as completed.
-func (m *manager) CompleteOngoingSwap(id types.Hash) {
+func (m *manager) CompleteOngoingSwap(id types.Hash) error {
 	m.Lock()
 	defer m.Unlock()
 	s, has := m.ongoing[id]
 	if !has {
-		return
+		return nil
```

Why is this a non-error, non-logging event. Do we sometimes call this function twice on the same ID?  Maybe we can add a comment for the answer?

---

good question, this should probably return an error since it shouldn't be called twice. will update